identity-account-storage/src/storage/memstore.rs

```diff
@@ -202,6 +206,85 @@ impl Storage for MemStore {
     }
   }
 
+  async fn key_exchange(
+    &self,
+    did: &IotaDID,
+    location: &KeyLocation,
+    public_key: PublicKey,
+    fragment: &str,
+  ) -> Result<KeyLocation> {
+    // Retrieves the PrivateKey from the vault
+    let mut vaults: RwLockWriteGuard<'_, _> = self.vaults.write()?;
+    let vault: &mut MemVault = vaults.get_mut(did).ok_or(Error::KeyVaultNotFound)?;
+    let key_pair: &KeyPair = vault.get(location).ok_or(Error::KeyNotFound)?;
+    if key_pair.type_() != KeyType::X25519 {
+      return Err(Error::InvalidKeyType(format!(
+        "{} expected KeyType::X25519",
+        key_pair.type_()
+      )));
+    }
+    // Computes the shared secret and inserts it in the vault
+    let shared_secret: [u8; 32] = X25519::key_exchange(key_pair.private(), public_key.as_ref())?;
+    let secret_location: KeyLocation =
+      KeyLocation::new(KeyType::X25519, fragment.to_owned(), key_pair.public().as_ref());
+    let secret_pair: KeyPair = KeyPair::from((
+      KeyType::X25519,
+      key_pair.public().clone(),
+      shared_secret.to_vec().into(),
+    ));
+    vault.insert(secret_location.clone(), secret_pair);
+    Ok(secret_location)
+  }
+
+  async fn encrypt_data(
+    &self,
+    did: &IotaDID,
+    location: &KeyLocation,
+    data: Vec<u8>,
+    associated_data: Vec<u8>,
+  ) -> Result<EncryptedData> {
+    // Retrieves the PrivateKey from the vault
+    let mut vaults: RwLockWriteGuard<'_, _> = self.vaults.write()?;
+    let vault: &mut MemVault = vaults.get_mut(did).ok_or(Error::KeyVaultNotFound)?;
+    let key_pair: &KeyPair = vault.get(location).ok_or(Error::KeyNotFound)?;
+    // Encrypts the given data
+    let mut cipher_text: Vec<u8> = vec![0; data.len()];
+    let mut tag: Vec<u8> = [0; Aes256Gcm::TAG_LENGTH].to_vec();
+    Aes256Gcm::try_encrypt(
+      key_pair.private().as_ref(),
+      [0; Aes256Gcm::NONCE_LENGTH].as_ref(),
```

Comment:
What should we do here? Nonce has to be the same when encrypting and decrypting the data

Reply:
I think these comments can be resolved. For the reader: `crypto.rs` is used to generate an appropriate nonce and that nonce is returned in `EncryptedData`.